sigstore/_verify/policy.py

```diff
@@ -0,0 +1,233 @@
+_OIDC_GITHUB_WORKFLOW_REF_OID = ObjectIdentifier("1.3.6.1.4.1.57264.1.6")
+_OTHERNAME_OID = ObjectIdentifier("1.3.6.1.4.1.57264.1.7")
+
+
+class _SingleX509ExtPolicy(ABC):
+    oid: ObjectIdentifier
+
+    def __init__(self, value: str) -> None:
+        self._value = value
+
+    def verify(self, cert: Certificate) -> VerificationResult:
+        try:
+            ext = cert.extensions.get_extension_for_oid(self.oid).value
+        except ExtensionNotFound:
+            return VerificationFailure(
+                reason=(
+                    f"Certificate does not contain {self.__class__.__name__} "
+                    f"({self.oid.dotted_string}) extension"
+                )
+            )
+
+        # NOTE(ww): mypy is confused by the `Extension[ExtensionType]` returned
+        # by `get_extension_for_oid` above.
+        ext_value = ext.value.decode()  # type: ignore[attr-defined]
+        if ext_value != self._value:
+            return VerificationFailure(
+                reason=(
+                    f"Certificate's {self.__class__.__name__} does not match "
+                    f"(got {ext_value}, expected {self._value})"
+                )
+            )
+
+        return VerificationSuccess()
+
+
+class OIDCIssuer(_SingleX509ExtPolicy):
+    """
+    Verifies the certificate's OIDC issuer, identified by
+    an X.509v3 extension tagged with `1.3.6.1.4.1.57264.1.1`.
+    """
+
+    oid = _OIDC_ISSUER_OID
+
+
+class GitHubWorkflowTrigger(_SingleX509ExtPolicy):
+    """
+    Verifies the certificate's GitHub Actions workflow trigger,
+    identified by an X.509v3 extension tagged with `1.3.6.1.4.1.57264.1.2`.
+    """
+
+    oid = _OIDC_GITHUB_WORKFLOW_TRIGGER_OID
+
+
+class GitHubWorkflowSHA(_SingleX509ExtPolicy):
+    """
+    Verifies the certificate's GitHub Actions workflow commit SHA,
+    identified by an X.509v3 extension tagged with `1.3.6.1.4.1.57264.1.3`.
+    """
+
+    oid = _OIDC_GITHUB_WORKFLOW_SHA_OID
+
+
+class GitHubWorkflowName(_SingleX509ExtPolicy):
+    """
+    Verifies the certificate's GitHub Actions workflow name,
+    identified by an X.509v3 extension tagged with `1.3.6.1.4.1.57264.1.4`.
+    """
+
+    oid = _OIDC_GITHUB_WORKFLOW_NAME_OID
+
+
+class GitHubWorkflowRepository(_SingleX509ExtPolicy):
+    """
+    Verifies the certificate's GitHub Actions workflow repository,
+    identified by an X.509v3 extension tagged with `1.3.6.1.4.1.57264.1.5`.
+    """
+
+    oid = _OIDC_GITHUB_WORKFLOW_REPOSITORY_OID
+
+
+class GitHubWorkflowRef(_SingleX509ExtPolicy):
+    """
+    Verifies the certificate's GitHub Actions workflow ref,
+    identified by an X.509v3 extension tagged with `1.3.6.1.4.1.57264.1.6`.
+    """
+
+    oid = _OIDC_GITHUB_WORKFLOW_REF_OID
+
+
+class VerificationPolicy(Protocol):
+    @abstractmethod
+    def verify(self, cert: Certificate) -> VerificationResult:
+        raise NotImplementedError  # pragma: no cover
+
+
+class AnyOf:
```

Comment:
On the subject of Protocol vs ABC I think there is a third option: keep using Protocol but still explicitly declare the subclass.

Currently it's not immediately obvious to reader that there are ~10 VerificationPolicy implementations in this file... I suppose more explicit naming might be another solution to that ( some policy classes are called *Policy, some derive from a *Policy, some do neither)